registration/include/pcl/registration/correspondence_estimation_lookup_table.h

```diff
@@ -0,0 +1,517 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id$
+ *
+ */
+
+#ifndef PCL_REGISTRATION_CORRESPONDENCE_ESTIMATION_LOOKUP_TABLE_H_
+#define PCL_REGISTRATION_CORRESPONDENCE_ESTIMATION_LOOKUP_TABLE_H_
+
+#include <algorithm>
+#include <cmath>
+#include <string>
+#include <vector>
+
+#include <Eigen/Core>
+
+#include <pcl/pcl_base.h>
+#include <pcl/pcl_macros.h>
+#include <pcl/common/common.h>
+#include <pcl/common/transforms.h>
+#include <pcl/search/search.h>
+#include <pcl/search/kdtree.h>
+#include <pcl/registration/correspondence_types.h>
+#include <pcl/registration/correspondence_estimation.h>
+
+namespace pcl
+{
+  namespace registration
+  {
+    /** \brief @b CorrespondenceLookupTableCell represents a match between
+      * a CorrespondenceLookupTable cell and the closest point in the reference point cloud.
+      *
+      * \author Carlos M. Costa
+      * \ingroup registration
+      */
+    struct CorrespondenceLookupTableCell
+    {
+      /** \brief Index of the closest point in the reference point cloud. */
+      int closest_point_index;
+
+      /** \brief Squared distance to the closest point. */
+      float distance_squared_to_closest_point;
+
+      /** \brief Empty constructor.
+        * Sets \ref closest_point_index to -1, \ref distance_to_closest_point to -FLT_MAX.
+        */
+      inline CorrespondenceLookupTableCell ()
+      : closest_point_index (-1), distance_squared_to_closest_point (-std::numeric_limits<float>::max ()) {}
+
+      /** \brief Constructor. */
+      inline CorrespondenceLookupTableCell (int index, float distance)
+      : closest_point_index (index), distance_squared_to_closest_point(distance) {}
+
+      /** \brief Empty destructor. */
+      virtual ~CorrespondenceLookupTableCell () {}
+    };
+
+    /** \brief @b LookupTable provides fast correspondence estimation
+      * by pre-computing the closest point for each cell within a uniform volume grid.
+      * It is recommended for 2D point cloud registration or 3D point cloud registration of
+      * small volumes (it requires more memory than k-d trees but it is much faster).
+      *
+      * \author Carlos M. Costa
+      * \ingroup registration
+      */
+    template <typename PointT>
+    class CorrespondenceLookupTable
+    {
+      public:
+        /** \brief Empty constructor. */
+        CorrespondenceLookupTable ()
+          : cell_resolution_ (0.01)
+          , cell_resolution_inverse_ (100.0)
+          , lookup_table_margin_ (1.0, 1.0, 1.0)
+          , number_cells_x_ (0)
+          , number_cells_y_ (0)
+          , number_cells_z_ (0)
+          , number_cells_xy_slice_ (0)
+          , use_search_tree_when_query_point_is_outside_lookup_table_ (true)
+          , compute_distance_from_query_point_to_closest_point_ (false)
+          , number_of_queries_on_lookup_table_ (0)
+          , number_of_queries_on_search_tree_ (0) {}
+
+        /** \brief Empty destructor. */
+        virtual ~CorrespondenceLookupTable () {}
```

Comment:
Self: This destructor is only justifiable if there's a good reason to have virtual methods.